### PR TITLE
Jetpack Backup: Refactor and unit test the Backup page date picker

### PR DESCRIPTION
**Pull request by mattgawarecki**
#### Changes proposed in this Pull Request

* Combine `client/components/jetpack/backup-date-picker` with `client/my-sites/backup/date-picker`, since the latter is only a wrapper for the former, and they're only used in the context of the Backup page.
* Convert the new `BackupDatePicker` from a class-based component to a functional one that takes advantage of React hooks.
* Convert the new `BackupDatePicker` from JavaScript to TypeScript.
* Add unit tests to cover `BackupDatePicker` and its hooks.

#### Testing instructions (regression)

On both Calypso Blue and Green:

1. Navigate to the Backup page.
2. Verify that you can navigate forward and backward, but only within the range of today (at the latest) to the oldest known backup attempt (at the earliest). _(For sites with Backup Daily, the earliest available date may likely be 30 days in the past.)_
3. Select a site whose timezone has an offset more negative than your browser's (e.g., if your browser uses UTC-6, set your site to use UTC-8). Verify you can navigate both forward and backward as normal, like in step 2. _(See #48269 for more details on this.)_
4. Manually alter your browser's URL to select a date that's sometime in the future. Verify that you can navigate backward toward the present, but not any further forward in time.
5. Do the same as in step 3, but select a date that's prior to the earliest known backup attempt. Verify that you can navigate forward, but not any further backward in time.

### CODE REVIEW

File: client/my-sites/backup/backup-date-picker/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+import { useLocalizedMoment } from 'calypso/components/localized-moment';
+import { useDateWithOffset, useFirstMatchingBackupAttempt } from '../hooks';
+
+export const useCanGoToDate = ( selectedDate: Moment, oldestDateAvailable?: Moment ) => {
+	const moment = useLocalizedMoment();
+	const today = useDateWithOffset( moment() );
+
+	return useCallback(
+		( desiredDate ) => {
```
Comment (rcanepa) [inline client/my-sites/backup/backup-date-picker/hooks.ts:23]: Could we add the `Moment` type to the `desiredDate` parameter and add `boolean` as the return type of the function? It's clear that the function signature should be `( desiredDate: Moment) => boolean` but TS can only infer that it's `( desiredDate: any) => any`.

File: client/my-sites/backup/backup-date-picker/index.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
+/**
```
Comment (rcanepa) [inline client/my-sites/backup/backup-date-picker/index.tsx:1]: This file looks very similar to `client/components/jetpack/backup-date-picker/index.jsx`. Is there a reason to have both?

File: client/my-sites/backup/main.jsx [hunk 3]
```diff
@@ -51,8 +51,16 @@ const BackupPage = ( { queryDate } ) => {
...
+	// If a date is specified, it'll be in a timezone-agnostic string format,
+	// so we'll need to add the site's TZ info in without affecting the date
+	// we were given.
+	//
+	// Otherwise, if no date is specified, we're talking about the current date.
+	// This is the same point in time for everyone, but we should make sure to
+	// store it in terms of the selected site's time zone.
 	const selectedDate = useDateWithOffset( parsedQueryDate, {
-		keepLocalTime: true,
+		keepLocalTime: !! queryDate,
```
Comment (rcanepa) [inline client/my-sites/backup/main.jsx:63]: Good catch!

**Review by rcanepa — APPROVED** [review]
Awesome work! Everything reads much more clearer now and things seem to work fine – I didn't find any regression.

Left some non-blocking comments. Great job!

File: client/my-sites/backup/backup-date-picker/index.tsx (hunk 2 above)
Comment (mattgawarecki) [reply client/my-sites/backup/backup-date-picker/index.tsx:1]: Nope, no need to keep both! I simply forgot to delete the `components/jetpack` version when I was done with it.

**Comment by rcanepa** [comment]
The latest changes look great. Awesome work!

:shipit:

**Review by rcanepa — APPROVED** [review]

**Pull request merged**